File: app/src/main/java/se/gustavkarlsson/skylight/android/observers/ObservableData.java
package se.gustavkarlsson.skylight.android.observers;

import java.util.IdentityHashMap;
import java.util.Set;

import static java.util.Collections.newSetFromMap;
import static java8.util.stream.StreamSupport.stream;

public class ObservableData<T> {
	private final Set<DataObserver<T>> listeners = newSetFromMap(new IdentityHashMap<>(20));
	private T data;

	public ObservableData(T data) {
		this.data = data;
	}

	public T getData() {
		return data;
	}

	public void setData(T data) {
		this.data = data;
		stream(listeners).forEach(listener -> listener.dataChanged(data));
	}

	public boolean addListener(DataObserver<T> listener) {
		return listeners.add(listener);
	}

	public boolean removeListener(DataObserver<T> listener) {
		return listeners.remove(listener);
	}

}

File: app/src/main/java/se/gustavkarlsson/skylight/android/observers/ObservableData.kt
package se.gustavkarlsson.skylight.android.observers

import java.util.*
import java.util.Collections.newSetFromMap

class ObservableData<T>(data: T) {
	var data: T = data
	set(data) {
        field = data
		listeners.forEach { it.dataChanged(data) }
    }

	private val listeners = newSetFromMap(IdentityHashMap<DataObserver<T>, Boolean>(20))

    fun addListener(listener: DataObserver<T>): Boolean {
        return listeners.add(listener)
    }

    fun removeListener(listener: DataObserver<T>): Boolean {
        return listeners.remove(listener)
    }

}
